runtime/src/snapshot_minimizer.rs

```diff
@@ -0,0 +1,609 @@
+        sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc, Mutex,
+        },
+    },
+};
+
+pub struct SnapshotMinimizer {
+    bank: Arc<Bank>,
+    starting_slot: Slot,
+    ending_slot: Slot,
+    minimized_account_set: DashSet<Pubkey>,
+}
+
+impl SnapshotMinimizer {
+    /// Removes all accounts not necessary for replaying slots in the range [starting_slot, ending_slot].
+    pub fn minimize(
+        bank: Arc<Bank>,
+        starting_slot: Slot,
+        ending_slot: Slot,
+        minimized_account_set: DashSet<Pubkey>,
+    ) {
+        let minimizer = Self::new(bank, starting_slot, ending_slot, minimized_account_set);
+
+        minimizer.add_accounts(
+            Self::get_rent_collection_accounts,
+            "rent collection accounts",
+        );
+        minimizer.add_accounts(Self::get_vote_accounts, "vote accounts");
+        minimizer.add_accounts(Self::get_stake_accounts, "stake accounts");
+        minimizer.add_accounts(Self::get_owner_accounts, "owner accounts");
+        minimizer.add_accounts(Self::get_programdata_accounts, "programdata accounts");
+
+        minimizer.minimize_accounts_db();
+
+        // Update accounts_cache and capitalization
+        minimizer.bank.force_flush_accounts_cache();
+        minimizer.bank.set_capitalization();
+    }
+
+    /// Create a minimizer - used in tests
+    fn new(
+        bank: Arc<Bank>,
+        starting_slot: Slot,
+        ending_slot: Slot,
+        minimized_account_set: DashSet<Pubkey>,
+    ) -> Self {
+        Self {
+            bank,
+            starting_slot,
+            ending_slot,
+            minimized_account_set,
+        }
+    }
+
+    /// Helper function to measure time and number of accounts added
+    fn add_accounts<F>(&self, add_accounts_fn: F, name: &'static str)
+    where
+        F: Fn(&SnapshotMinimizer),
+    {
+        let initial_accounts_len = self.minimized_account_set.len();
+        let (_, measure) = measure!(add_accounts_fn(self), name);
+        let total_accounts_len = self.minimized_account_set.len();
+        let added_accounts = total_accounts_len - initial_accounts_len;
+
+        info!(
+            "Added {added_accounts} {name} for total of {total_accounts_len} accounts. get {measure}"
+        );
+    }
+
+    /// Used to get rent collection accounts in `minimize`
+    fn get_rent_collection_accounts(&self) {
+        let partitions = if !self.bank.use_fixed_collection_cycle() {
+            self.bank
+                .variable_cycle_partitions_between_slots(self.starting_slot, self.ending_slot)
+        } else {
+            self.bank
+                .fixed_cycle_partitions_between_slots(self.starting_slot, self.ending_slot)
+        };
+
+        partitions.into_iter().for_each(|partition| {
+            let subrange = Bank::pubkey_range_from_partition(partition);
+            self.bank
+                .accounts()
+                .load_to_collect_rent_eagerly(&self.bank.ancestors, subrange)
```

Comment:
this looks like it will work.
all you need is the pubkeys from the scan. You don't need to load accounts.
could take the guts from here:
```


    pub fn load_to_collect_rent_eagerly<R: RangeBounds<Pubkey> + std::fmt::Debug>(
        &self,
        ancestors: &Ancestors,
        range: R,
    ) -> Vec<PubkeyAccountSlot> {
        self.accounts_db.range_scan_accounts(
            "", // disable logging of this. We now parallelize it and this results in multiple parallel logs
            ancestors,
            range,
            &ScanConfig::new(true),
            |collector: &mut Vec<PubkeyAccountSlot>, option| {
               < use option.pubkey to insert pubkeys>
            },
        )
    }
    
    ```
But, what you did is possibly less code and should still work. And your feature is just ledger-tool.

Maybe comment here that we could use the guts instead So probably ok to leave this. But, if this were re-used in the runtime somehow, this is overkill.